File: grpc/grpc-client-annotation-processor/src/main/java/ru/tinkoff/grpc/client/annotation/processor/GrpcClientExtension.java
package ru.tinkoff.grpc.client.annotation.processor;

import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import ru.tinkoff.kora.annotation.processor.common.AnnotationUtils;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionResult;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.Nullable;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.*;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import static ru.tinkoff.grpc.client.annotation.processor.GrpcClassNames.*;

public final class GrpcClientExtension implements KoraExtension {
    private final ProcessingEnvironment env;
    private final TypeElement stubTypeElement;
    private final TypeMirror stubErasure;

    public GrpcClientExtension(ProcessingEnvironment env, TypeElement stubTypeElement, TypeMirror stubErasure) {
        this.env = env;
        this.stubTypeElement = stubTypeElement;
        this.stubErasure = stubErasure;
    }

    @Nullable
    @Override
    public KoraExtensionDependencyGenerator getDependencyGenerator(RoundEnvironment roundEnvironment, TypeMirror typeMirror, Set<String> tag) {
        if (stubTypeElement == null || stubErasure == null) {
            return null;
        }
        if (typeMirror.getKind() != TypeKind.DECLARED) {
            return null;
        }
        var dt = (DeclaredType) typeMirror;
        var te = (TypeElement) dt.asElement();
        var className = ClassName.get(te);
        if (env.getTypeUtils().isAssignable(typeMirror, stubErasure)) {
            return getStubGenerator(typeMirror, tag);
        }

        if (className.equals(channel) && tag.size() == 1) {
            return getChannel(typeMirror, tag);
        }
        if (className.equals(serviceDescriptor) && tag.size() == 1) {
            return getServiceDescriptor(typeMirror, tag);
        }

        return null;
    }

    private KoraExtensionDependencyGenerator getServiceDescriptor(TypeMirror typeMirror, Set<String> tag) {
        var serviceDescriptorTypeElement = env.getElementUtils().getTypeElement(serviceDescriptor.canonicalName());
        var serviceDescriptorTypeMirror = serviceDescriptorTypeElement.asType();

        var grpcServiceTypeElement = env.getElementUtils().getTypeElement(tag.iterator().next());
        var grpcServiceClassName = ClassName.get(grpcServiceTypeElement);

        return () -> new ExtensionResult.CodeBlockResult(
            grpcServiceTypeElement,
            params -> CodeBlock.of("$T.getServiceDescriptor()", grpcServiceClassName),
            serviceDescriptorTypeMirror,
            tag,
            List.of(),
            List.of()
        );
    }

    private KoraExtensionDependencyGenerator getChannel(TypeMirror typeMirror, Set<String> tag) {
        var managedChannelTypeElement = env.getElementUtils().getTypeElement(managedChannelLifecycle.canonicalName());
        var managedChannelTypeMirror = managedChannelTypeElement.asType();
        var managedChannelConstructor = CommonUtils.findConstructors(managedChannelTypeElement, m -> m.contains(Modifier.PUBLIC)).get(0);
        var parameterTags = new ArrayList<Set<String>>(managedChannelConstructor.getParameters().size());
        var parameterTypes = new ArrayList<TypeMirror>(managedChannelConstructor.getParameters().size());
        for (int i = 0; i < managedChannelConstructor.getParameters().size(); i++) {
            var parameter = managedChannelConstructor.getParameters().get(i);
            parameterTags.add(i < 3 ? tag : Set.of());
            parameterTypes.add(parameter.asType());
        }
        return () -> new ExtensionResult.CodeBlockResult(
            managedChannelConstructor,
            params -> CodeBlock.of("new $T($L)", managedChannelLifecycle, params),
            managedChannelTypeMirror,
            tag,
            parameterTypes,
            parameterTags
        );
    }

    @Nullable
    private KoraExtensionDependencyGenerator getStubGenerator(TypeMirror typeMirror, Set<String> tag) {
        if (!tag.isEmpty()) {
            return null;
        }
        var typeElement = env.getTypeUtils().asElement(typeMirror);
        var apiTypeElement = typeElement.getEnclosingElement();
        if (apiTypeElement.getKind() != ElementKind.CLASS) {
            return null;
        }
        var apiClassName = ClassName.get((TypeElement) apiTypeElement);
        if (AnnotationUtils.findAnnotation(apiTypeElement, grpcGenerated) == null) {
            return null;
        }
        var typeName = typeMirror.toString();
        final ExecutableElement sourceElement;
        if (typeName.endsWith("BlockingStub")) {
            sourceElement = this.findMethod(apiTypeElement, "newBlockingStub");
        } else if (typeName.endsWith("FutureStub")) {
            sourceElement = this.findMethod(apiTypeElement, "newFutureStub");
        } else {
            sourceElement = this.findMethod(apiTypeElement, "newStub");
        }
        var channelType = sourceElement.getParameters().get(0).asType();

        return () -> new ExtensionResult.CodeBlockResult(
            sourceElement,
            params -> CodeBlock.of("$T.$N($L)", apiClassName, sourceElement.getSimpleName(), params),
            typeMirror,
            tag,
            List.of(channelType),
            List.of(Set.of(apiClassName.canonicalName()))
        );
    }

    private ExecutableElement findMethod(Element apiTypeElement, String methodName) {
        for (var enclosedElement : apiTypeElement.getEnclosedElements()) {
            if (enclosedElement.getKind() != ElementKind.METHOD) {
                continue;
            }
            if (!enclosedElement.getModifiers().contains(Modifier.STATIC)) {
                continue;
            }
            if (!enclosedElement.getModifiers().contains(Modifier.PUBLIC)) {
                continue;
            }
            if (enclosedElement.getSimpleName().contentEquals(methodName)) {
                return (ExecutableElement) enclosedElement;
            }
        }
        throw new IllegalStateException();
    }
}

File: grpc/grpc-client-ksp/src/main/kotlin/ru.tinkoff.grpc.client.ksp/GrpcClientExtension.kt
package ru.tinkoff.grpc.client.ksp

import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.getConstructors
import com.google.devtools.ksp.getDeclaredFunctions
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.grpc.client.ksp.GrpcClassNames.channel
import ru.tinkoff.grpc.client.ksp.GrpcClassNames.grpcGenerated
import ru.tinkoff.grpc.client.ksp.GrpcClassNames.managedChannelLifecycle
import ru.tinkoff.grpc.client.ksp.GrpcClassNames.serviceDescriptor
import ru.tinkoff.grpc.client.ksp.GrpcClassNames.stubFor
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionResult.CodeBlockResult
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findAnnotation
import ru.tinkoff.kora.ksp.common.AnnotationUtils.findValueNoDefault
import ru.tinkoff.kora.ksp.common.AnnotationUtils.isAnnotationPresent
import ru.tinkoff.kora.ksp.common.KspCommonUtils.getClassDeclarationByName

class GrpcClientExtension(
    val resolver: Resolver,
    val kspLogger: KSPLogger,
    val codeGenerator: CodeGenerator,
    val abstractStub: KSClassDeclaration?,
    val abstractCoroutineStub: KSClassDeclaration?) : KoraExtension {

    override fun getDependencyGenerator(resolver: Resolver, type: KSType, tag: Set<String>): (() -> ExtensionResult)? {
        if (abstractStub != null && abstractStub.asStarProjectedType().isAssignableFrom(type)) {
            return this.generateJavaStub(type, tag)
        }
        if (abstractCoroutineStub != null && abstractCoroutineStub.asStarProjectedType().isAssignableFrom(type)) {
            return this.generateCoroutineStub(type, tag)
        }
        val typeName = type.toTypeName()
        if (typeName == channel) {
            return this.generateChannel(type, tag)
        }
        if (typeName == serviceDescriptor) {
            return this.generateGrpcServiceDescriptor(type, tag)
        }
        return null
    }

    private fun generateJavaStub(type: KSType, tag: Set<String>): (() -> ExtensionResult)? {
        if (tag.isNotEmpty()) {
            return null
        }
        val classDecl = type.declaration
        if (classDecl !is KSClassDeclaration) {
            return null
        }
        val apiClassDecl = classDecl.parentDeclaration
        if (apiClassDecl !is KSClassDeclaration) {
            return null
        }
        val apiClassName = apiClassDecl.toClassName()
        if (!apiClassDecl.isAnnotationPresent(grpcGenerated)) {
            return null
        }
        val typeName = classDecl.simpleName.asString()
        val sourceElement = when {
            typeName.endsWith("BlockingStub") -> this.findMethod(apiClassDecl, "newBlockingStub")
            typeName.endsWith("FutureStub") -> this.findMethod(apiClassDecl, "newFutureStub")
            else -> this.findMethod(apiClassDecl, "newStub")
        }
        val channelType = sourceElement.parameters[0].type.resolve()
        return {
            CodeBlockResult(
                sourceElement,
                { CodeBlock.of("%T.%N(%L)", apiClassName, sourceElement.simpleName.asString(), it) },
                type,
                tag,
                listOf(channelType),
                listOf(setOf(apiClassName.canonicalName))
            )
        }
    }

    private fun findMethod(apiClassDecl: KSClassDeclaration, name: String) = apiClassDecl.getDeclaredFunctions()
        .filter { it.simpleName.asString() == name }
        .first()

    private fun generateGrpcServiceDescriptor(type: KSType, tag: Set<String>): (() -> ExtensionResult)? {
        if (tag.isEmpty()) {
            return null
        }
        val serviceDescriptorDecl = resolver.getClassDeclarationByName(serviceDescriptor)!!
        val serviceDescriptorType = serviceDescriptorDecl.asStarProjectedType()

        val grpcServiceDecl = resolver.getClassDeclarationByName(tag.first())!!
        val grpcServiceClassName = grpcServiceDecl.toClassName()
        return {
            CodeBlockResult(
                grpcServiceDecl,
                { CodeBlock.of("%T.getServiceDescriptor()", grpcServiceClassName) },
                serviceDescriptorType,
                tag,
                listOf(),
                listOf()
            )
        }
    }

    private fun generateChannel(type: KSType, tag: Set<String>): (() -> ExtensionResult)? {
        if (tag.isEmpty()) {
            return null
        }
        val managedChannelDecl = resolver.getClassDeclarationByName(managedChannelLifecycle.canonicalName)!!
        val managedChannelType = managedChannelDecl.asStarProjectedType()
        val managedChannelConstructor = managedChannelDecl.getConstructors().first()
        val parameterTags = arrayListOf<Set<String>>()
        val parameterTypes = arrayListOf<KSType>()
        for ((i, parameter) in managedChannelConstructor.parameters.withIndex()) {
            if (i < 3) {
                parameterTags.add(tag)
            } else {
                parameterTags.add(setOf())
            }
            parameterTypes.add(parameter.type.resolve())
        }
        return {
            CodeBlockResult(
                managedChannelConstructor,
                { CodeBlock.of("%T(%L)", managedChannelLifecycle, it) },
                managedChannelType,
                tag,
                parameterTypes,
                parameterTags
            )
        }
    }


    private fun generateCoroutineStub(type: KSType, tag: Set<String>): (() -> ExtensionResult)? {
        if (tag.isNotEmpty()) {
            return null
        }
        val classDecl = type.declaration
        if (classDecl !is KSClassDeclaration) {
            return null
        }
        val apiClassDecl = classDecl.parentDeclaration
        if (apiClassDecl !is KSClassDeclaration) {
            return null
        }
        val stubForAnnotation = apiClassDecl.findAnnotation(stubFor)
        if (stubForAnnotation == null) {
            return null
        }
        val apiTag = stubForAnnotation.findValueNoDefault<List<KSType>>("value")!!
            .map { it.toClassName().canonicalName }
            .toSet()
        val implClassName = classDecl.toClassName()
        val constructor = classDecl.primaryConstructor!!
        val channelType = constructor.parameters[0].type.resolve()
        return {
            CodeBlockResult(
                classDecl,
                { CodeBlock.of("%T(%L)", implClassName, it) },
                type,
                apiTag,
                listOf(channelType),
                listOf(apiTag)
            )
        }
    }
}

